stream_alert/shared/publisher.py

```diff
@@ -0,0 +1,272 @@
+
+        [output service name].[field name]
+
+        E.g. "demisto.blah"
+
+        Args:
+            alert (Alert): The alert instance to publish.
+            publication (dict): An existing publication generated by previous publishers in the
+                series of publishers, or {}.
+
+        Returns:
+            dict: The published alert.
+        """
+
+
+class CompositePublisher(AlertPublisher):
+    """A publisher class that combines the logic of multiple other publishers together in series
+
+    To reduce the chance that one publisher has side effects in other publishers in the chain,
+    we use deepcopy between the publishers.
+
+    Note: This publisher is not meant to be @Register'd as it does not have any logic on its own.
+          It is only meant to be composed by AlertPublisherRepository to give a common interface to
+          multiple publishers chained in sequence.
+    """
+
+    def __init__(self, publishers):
+        self._publishers = publishers  # Type list(BaseAlertPublisher)
+
+        for publisher in self._publishers:
+            if not isinstance(publisher, AlertPublisher):
+                LOGGER.error('CompositePublisher given invalid publisher')
+
+    def publish(self, alert, publication):
+        for publisher in self._publishers:
+            try:
+                publication = deepcopy(publication)
+                publication = publisher.publish(alert, publication)
+            except KeyError:
+                LOGGER.exception(
+                    'CompositePublisher encountered KeyError with publisher: %s',
+                    publisher.__name__
+                )
+                raise
+
+        return publication
+
+
+class WrappedFunctionPublisher(AlertPublisher):
+    """A class only used to wrap a function publisher."""
+
+    def __init__(self, function):
+        self._function = function
+
+    def publish(self, alert, publication):
+        return self._function(alert, publication)
+
+
+class AlertPublisherRepository(object):
+    """A repository mapping names -> publishers
+
+    As a usability optimization, using this Repository will eagerly load and register all
+    publishers in the application.
+    """
+    _publishers = {}
+
+    @staticmethod
+    def is_valid_publisher(thing):
+        """Returns TRUE if the given reference can be registered as a publisher
+
+        Publishers are valid if and only if they fall into one of the following categories:
+
+        * They are a python function that accepts 2 arguments: (alert: Alert, publication: dict)
+        * They are a python class that extends AlertPublisher
+
+        Args:
+            thing (mixed): Any primitive or reference to be checked
+
+        Returns:
+            bool
+        """
+
+        # We have to put the isclass() check BEFORE the callable() check because classes are also
+        # callable!
+        if isclass(thing) and issubclass(thing, AlertPublisher):
```

Comment:
this could be a oneliner (if you'd want):

`return (isclass(thing) and issubclass(thing, AlertPublisher)) or callable(thing)`